src/lib/genesis_ledger_helper/genesis_ledger_helper.ml

```diff
@@ -46,110 +50,420 @@ module Tar = struct
+  let find_tar ~logger ~constraint_constants (config : Runtime_config.Ledger.t)
+      =
+    let search_paths =
+      Cache_dir.
+        [autogen_path; manual_install_path; brew_install_path; s3_install_path]
+    in
+    let file_exists filename path =
+      let filename = path ^/ filename in
+      if%map file_exists ~follow_symlinks:true filename then (
+        Logger.info ~module_:__MODULE__ ~location:__LOC__ logger
+          "Found ledger file at $path"
+          ~metadata:[("path", `String filename)] ;
+        Some filename )
+      else (
+        Logger.info ~module_:__MODULE__ ~location:__LOC__ logger
+          "Ledger file $path does not exist"
+          ~metadata:[("path", `String filename)] ;
+        None )
+    in
+    let%bind hash_filename =
+      match config.hash with
+      | Some hash -> (
+          let hash_filename = hash_filename hash in
+          let%bind tar_path =
+            Deferred.List.find_map ~f:(file_exists hash_filename) search_paths
           in
-          (sk, account) )
+          match tar_path with
+          | Some _ ->
+              return tar_path
+          | None -> (
+              let s3_path = s3_bucket_prefix ^/ hash_filename in
+              let local_path = Cache_dir.s3_install_path ^/ hash_filename in
+              match%bind
+                Cache_dir.load_from_s3 [s3_path] [local_path] ~logger
+              with
+              | Ok () ->
+                  file_exists hash_filename Cache_dir.s3_install_path
+              | Error e ->
+                  Logger.info ~module_:__MODULE__ ~location:__LOC__ logger
+                    "Could not download genesis ledger from $uri: $error"
+                    ~metadata:
+                      [ ("uri", `String s3_path)
+                      ; ("error", `String (Error.to_string_hum e)) ] ;
+                  return None ) )
+      | None ->
+          return None
     in
+    match hash_filename with
+    | Some filename ->
+        return (Some filename)
+    | None -> (
+      match config.base with
+      | Hash hash ->
+          assert (Some hash = config.hash) ;
+          return None
+      | Accounts (_accounts, None) ->
+          return None
+      | Accounts (_accounts, Some {name; newer_than}) ->
+          let newer_than =
+            Genesis_constants.validate_time (Some newer_than)
+            |> Result.map_error ~f:(fun str ->
+                   Error.of_string ("Could not parse 'newer_than': " ^ str) )
+            |> Or_error.ok_exn
+          in
+          let named_filename =
+            named_filename ~constraint_constants
+              ~num_accounts:config.num_accounts ("custom_" ^ name)
+          in
+          Deferred.List.find_map search_paths ~f:(fun path ->
+              match%bind file_exists named_filename path with
+              | None ->
+                  return None
+              | Some filename ->
+                  let%map {mtime; _} = Unix.stat filename in
```

Comment:
Relying on timestamps makes builds less reproducible -- can we instead do something that computes a hash? Proposal: Drop the ledger.hash field from the schema, hash the contents of the `ledger` sub-object of the config JSON, name the file that hash. And then we'll cache hit based off of that hash. That way if you want to change the genesis ledger, you just change the contents of the ledger in your config and it will hash to something else when you restart.

Reply:
This is an adequate solution and would achieve the same behavior we're doing w/ the timestamp I believe.

Reply:
This is done.